lib/molinillo/resolution.rb

```diff
@@ -831,9 +835,14 @@ def handle_missing_or_push_dependency_state(state)
           state.activated.detach_vertex_named(state.name)
           push_state_for_requirements(state.requirements.dup, false, state.activated)
         else
-          states.push(state).tap { activated.tag(state) }
+          push_state(state)
         end
       end
+
+      def push_state(state)
+        states_by_requirement[state.requirement] = state
```

Comment:
I think this needs to be `||=`, and also popping needs to remove / update the hash